### ISSUE TITLE
Add-on does not install when a Firefox update is staged

### ISSUE THREAD

**Issue by ckprice** [mozilla/testpilot#714]
I _think_ I've narrowed this odd ball down.

STR:
- Have an update staged (I was using Nightly).
- Visit https://testpilot.firefox.com.
- Click on `Install add-on`.

Results:
- Add-on dialogue proceeds normally, but page stays spinning.
- There is no icon in the toolbar.
- Refreshing page prompts me with the `Install add-on` button again.
- Visiting `about:addons` I see that Test Pilot looks to have installed.
- Restarting fixes things.

![bug](https://s3.amazonaws.com/f.cl.ly/items/0k291r2J0w2i3s1U2g41/Image%202016-05-05%20at%207.37.59%20AM.png?v=b042b197)

**Comment by meandavejustice** [comment]
What a weird edge case. Think I'll need to ask around in `#addons` or `#amo` to get to the bottom of this.

**Comment by pdehaan** [comment]
I can reproduce this with my staged FF Nightly install on OS X. Here's my Browser Console, but I can't find anything too interesting. Clicking the "testpilot-addon: Object" results brings up the sidebar, but I'm not sure how the output Object is relevant.

![firefox_test_pilot_and_browser_console_and__mozilla_testpilot__add-on_does_not_install_when_a_firefox_update_is_staged___714__-_peterdehaan_gmail_com_-_gmail](https://cloud.githubusercontent.com/assets/557895/15049380/e1d28628-12a2-11e6-9cd2-3ab882c573a9.png)

---

https://testpilot.firefox.com/__version__ (ca8af43986ea2e0f8f8ecb08625b26f1aae7a6e5): 

``` json
{ "build": "https://circleci.com/gh/mozilla/testpilot/325",
  "source": "https://github.com/mozilla/testpilot",
  "version": "2016-05-04",
  "commit": "ca8af43986ea2e0f8f8ecb08625b26f1aae7a6e5" }
```

**Comment by edwindotcom** [comment]
note: 'Staged' means "browser has notified and waiting for the user to restart to update Fx version"

**Comment by pdehaan** [comment]
Also, FWIW, it appears that the add-on is installing for me, and pointing to the correct env (Production), just not completing the flow.
![add-ons_manager](https://cloud.githubusercontent.com/assets/557895/15049536/a016b410-12a3-11e6-84ee-a3dea34e8b85.png)

**Comment by lmorchard** [comment]
> Clicking the "testpilot-addon: Object" results brings up the sidebar, but I'm not sure how the output Object is relevant.

Error reporting in the add-on is terrible, <strike>sometimes</strike> most of the time. If you click on that "object" message, somewhere buried in the sidebar is probably the actual text of the error message. You might have to drill down into a few object properties to get to the traceback there. Seems like we _might_ have something in there that we can fix during add-on init, but I'm having trouble reproducing this issue myself.

**Comment by meandavejustice** [comment]
@pdehaan If you can just paste the values of the 'fileName', 'stack', and 'message' keys in an issue comment or gist I can probably decipher from that. They are ugly but the messages are buried in there. This would be super helpful since this is a hard one to reproduce.

**Comment by pdehaan** [comment]
@meandavejustice I'm totally lost.
Clicking the "5 more..." link gives me the following sidebar with [seemingly unrelated] properties and `fileName` and `lineNumber`.

<img width="991" alt="browser_console" src="https://cloud.githubusercontent.com/assets/557895/15056872/c959be56-12c8-11e6-8889-339976055897.png">
1. runner.js:166 points to view-source:resource://gre/modules/commonjs/sdk/addon/runner.js
   
   ``` js
   } catch (error) {
     console.exception(error);
     throw error;
   }
   ```
2. Promise-backend.js:940 points to view-source:resource://gre/modules/Promise-backend.js
   
   ``` js
     } else if (typeof(this.onReject) == "function") {
         nextValue = this.onReject.call(undefined, nextValue);
         nextStatus = STATUS_RESOLVED;
     }
   ```

I can't seem to get a usable stack out of either result to see where the root cause may be.

**Comment by lmorchard** [comment]
The runner & promise-backend parts are the terrible red herrings, because they're just the end points where errors in Promises eventually propagate to if nothing else handles or reports them.

The important part is in the sidebar, about "message: shouldLog is not a function". I think I've seen that before, probably a month or two ago, but no idea what causes it - it's actually an error from somewhere in Firefox when something attempts to use `console.log` or similar and something is broken in there... somehow? 

I haven't been able to reproduce the conditions, but when that error starts showing up _all_ attempts to call `console.log` produce the error. Seems to go away when I restart Firefox.

**Comment by jaredhirsch** [comment]
`shouldLog is not a function` is thrown if you try to use `Console.jsm` after that module has been unloaded. I've seen it when I mess up the `Cu.unload` sequence in the search add-on.

If you're seeing this in devtools, I kinda wonder if it's a devtools bug?

At any rate, it sounds like we should detect if an update is pending on restart, and warn the user that they should restart _before_ installing stuff. I can look through dxr for some kind of state we can check.

**Comment by jaredhirsch** [comment]
Looks like the `nsUpdateService` fires an `update-staged` nsIObserver signal when an update is staged and ready for restart. It [looks like](https://dxr.mozilla.org/mozilla-central/source/toolkit/mozapps/update/nsUpdateService.js#3064) that signal also includes a flag to tell you the status of the update. We could listen for that signal and show the user some warning UI.

[Here's](https://dxr.mozilla.org/mozilla-central/source/toolkit/mozapps/update/nsUpdateService.js#79-89) a complete list of flags; thankfully they are strings, rather than integer flags. Searching around DXR should explain which ones matter. I'm guessing `STATE_APPLIED` (which maps to the string `"applied"` is the one we really care about.

I can look more after public launch, unless this is a blocker or super duper nice to have, in which case, I'll look after I finish the duplicate checking bug for search.

**Comment by jaredhirsch** [comment]
Hmm, I guess we'll need to detect this update pending state when the addon isn't installed at the time the `update-staged` signal is fired.

I think the best we could do would be to check some Gecko state when the addon is installed; if the update pending state is set, notify the user that stuff might look weird until they restart.

I looked a bit more through DXR but don't immediately see where that state is stored. There are a couple of update services that we can try, though. I bet we can find something.

**Comment by johngruen** [comment]
we are happy

**Comment by ghost** [comment]
Just tested in nightly and it works for me.
